Follow untyped imports

mcocdawc opened the pull request:
we now follow untyped imports.
   
Before this change anything from an untyped module was replaced by `Any`.
Even declaring an argument of a function as `pyscf.gto.mole.Mole` actually did not register its type as `pyscf.gto.mole.Mole` but as `Any`. :-(

mcocdawc commented:
@mscho527 @lweisburn 

Cranking up the `mypy` checking revealed the following potential bug

```text
src/quemb/molbe/lo.py:255: error: Item "EdmistonRuedenberg" of "EdmistonRuedenberg | PipekMezey | Boys" has no attribute "pop_method"  [union-attr]
src/quemb/molbe/lo.py:255: error: Item "Boys" of "EdmistonRuedenberg | PipekMezey | Boys" has no attribute "pop_method"  [union-attr]
src/quemb/kbe/pbe.py:150: error: Unexpected keyword argument "dm_kpts" for "get_veff"  [call-arg]
/home/mcocdawc/.venv/lib/python3.12/site-packages/pyscf/pbc/scf/hf.py:697: note: "get_veff" defined here
src/quemb/kbe/pbe.py:150: error: Unexpected keyword argument "kpts" for "get_veff"; did you mean "kpt"?  [call-arg]
/home/mcocdawc/.venv/lib/python3.12/site-packages/pyscf/pbc/scf/hf.py:697: note: "get_veff" defined here
src/quemb/kbe/pbe.py:227: error: Unexpected keyword argument "dm_kpts" for "get_veff"  [call-arg]
/home/mcocdawc/.venv/lib/python3.12/site-packages/pyscf/pbc/scf/hf.py:697: note: "get_veff" defined here
src/quemb/kbe/pbe.py:227: error: Unexpected keyword argument "kpts" for "get_veff"; did you mean "kpt"?  [call-arg]
/home/mcocdawc/.venv/lib/python3.12/site-packages/pyscf/pbc/scf/hf.py:697: note: "get_veff" defined here
```

lweisburn commented:
> @mscho527 @lweisburn
> 
> Cranking up the `mypy` checking revealed the following potential bug
> 
> ```
> src/quemb/molbe/lo.py:255: error: Item "EdmistonRuedenberg" of "EdmistonRuedenberg | PipekMezey | Boys" has no attribute "pop_method"  [union-attr]
> src/quemb/molbe/lo.py:255: error: Item "Boys" of "EdmistonRuedenberg | PipekMezey | Boys" has no attribute "pop_method"  [union-attr]
> ```
> 
> I ignored it for the time being. But is this a proper bug?

I don't know if this is considered a "proper bug", but the `pop_method` attribute only matters if you use PM localization. It doesn't exist for the other localization methods. It is None by default in the code, and I doubt many people would change this unless they really know what they're doing and want to choose a specialized localization. If you specified a population method and didn't choose PM localization, the calculation would crash and give some descriptive pyscf errors. (Like we talked about) there are lots of ways to change and fix this!

mcocdawc commented:
Thank you for the clarification @lweisburn 

I changed it in the function to
```python3
if pop_method is not None:
    assert isinstance(Localizer, PipekMezey)
    mlo.pop_method = pop_method
```
and added two overloads so that is explicit that for everything that is not PipekMezey the pop_method has to be None.

This makes it explicit to the human and to the type-checker.

mscho527 commented on tests/chem_dm_kBE_test.py:
```diff
@@ -9,16 +9,12 @@
 import os
 import unittest
 
+import libdmet
```
If we are always importing `libdmet` now, these decorators in the rest of this file below can go away, right?
```
@unittest.skipIf(libdmet is None, "Module `libdmet` not imported correctly.")
```

mcocdawc replied:
good catch thank you!

mscho527 reviewed (changes requested):
looks good! some minor questions & comments :)

mcocdawc commented on src/quemb/kbe/pbe.py:
```diff
@@ -50,7 +50,7 @@ class BE(Mixin_k_Localize):
     def __init__(
         self,
-        mf: pbc.scf.hf.SCF,
+        mf: scf.khf.KSCF,
```
thank you for pointing this out. I changed it.

mscho527 commented on mypy.ini:
```diff
@@ -2,6 +2,7 @@
     no_implicit_optional = True
     disallow_untyped_defs = True
     check_untyped_defs = True
+    follow_untyped_imports = True
```
sure cool! as long as we are not making the checks take too long & they don't hinder with feature additions, i am good with the typing additions.

mcocdawc commented on src/quemb/molbe/lo.py:
```diff
@@ -230,24 +235,46 @@ def get_pao_native(
     return Cpao
 
 
+@overload
+def get_loc(
+    mol: Mole,
+    C: Matrix,
+    method: Literal["PM"],
+    pop_method: str | None = ...,
+    init_guess: Matrix | None = ...,
+) -> Mole: ...
+
+
+@overload
+def get_loc(
+    mol: Mole,
+    C: Matrix,
+    method: Literal["ER", "FB"],
+    pop_method: None = ...,
+    init_guess: Matrix | None = ...,
+) -> Mole: ...
+
+
```
hmm, I would rather not do it. 

It is a new python feature yes, but it is in the standard library and not expected to change in the future. 
After one gets the gist of it, it behaves quite "regular" as in unsurprising for new developers.

In a few months it will be normal for us to see `@overload` in the code and comments explaining it would seem like unnecessary clutter. Just for comparison: we also don't put comments to explain the difference between a `@classmethod` and a `@staticmethod`, although they are also a non-trivial concept if you never heard about them.

Nevertheless, I think it is a good idea to discuss the `@overload` decorator and other more advanced mypy topics such as type narrowing in a group seminar.

mscho527 replied:
Yeah, we should probably discuss this as a group.

I personally think that normalizing function overloads start to walk the lines between using type checkers as an easy and useful check vs. defining a particular way of writing the code. My request for a comment explaining `@overload` was based on an assumption that we won't use it often in the code, but if we plan to, this would be worth a discussion.
